src/util/strutil.c

```diff
@@ -8,65 +8,132 @@
 #include <ctype.h>
 #include "RG.h"
 #include "rmalloc.h"
-#include "deps/utf8proc/utf8proc.h"
+#include "utf8proc/utf8proc.h"
 
-int str_length(const char *str) {
-	size_t str_len = strlen(str);
-	utf8proc_int32_t c;
-	size_t i = 0;
-	int len = 0;
-	while(i < str_len) {
-		i+= utf8proc_iterate(str + i, -1, &c);
-		len++;
-	}
-	return len;
-}
-
-void str_tolower(const char *str, char *lower, size_t *lower_len) {
+void str_tolower_ascii
+(
+	const char *str,
+	char *lower,
+	size_t *lower_len
+) {
 	size_t str_len = strlen(str);
-	//Avoid overflow
+	// avoid overflow
 	ASSERT(*lower_len >= str_len);
 
-	//Update lower len
+	// update lower len
 	*lower_len = str_len;
 
+	for(size_t i = 0; i < str_len; i++) lower[i] = tolower(str[i]);
+	lower[str_len] = 0;
+}
+
+bool str_utf8_validate
+(
+    const char *str
+) {
+	// hold current Unicode character
+    utf8proc_int32_t c;
+
+    // while we didn't get to the end of the string
+    while(str[0] != 0) {
+        // increment current position by number of bytes in Unicode character
+        str += utf8proc_iterate((const utf8proc_uint8_t *)str, -1, &c);
+		if(c == -1) return false;
+    }
+
+    // return the length of the string in terms of Unicode characters
+    return true;
+}
+
+// determine the length of the string
+// not in terms of the number of bytes in the string
+// but in terms of the number of Unicode characters in the string
+int str_length
+(
+    const char *str
+) {
+	// hold current Unicode character
+    utf8proc_int32_t c;
+
+    int len = 0;
+
+    // while we didn't get to the end of the string
+    while(str[0] != 0) {
+        // increment current position by number of bytes in Unicode character
+        str += utf8proc_iterate((const utf8proc_uint8_t *)str, -1, &c);
+        // increment length of the string in terms of Unicode characters
+        len++;
+    }
+
+    // return the length of the string in terms of Unicode characters
+    return len;
+}
+
+char *str_tolower
+(
+	const char *str
+) {
+	// hold current Unicode character
 	utf8proc_int32_t c;
-	size_t i = 0;
-	int w = 0;
-	while(i < str_len) {
-		w = utf8proc_iterate(str + i, -1, &c);
-		utf8proc_encode_char(utf8proc_tolower(c), lower + i);
-		i+=w;
+	const char *str_i = str;
+	size_t lower_len = 0;
+	char buffer[4];
+	while(str_i[0] != 0) {
+		str_i     += utf8proc_iterate((const utf8proc_uint8_t *)str_i, -1, &c);
+		lower_len += utf8proc_encode_char(utf8proc_tolower(c), (utf8proc_uint8_t *)buffer);
```

Comment:
are we expecting an upper case unicode character and its corresponding lower case counter part to differ in size?

Reply:
yes